SC_redesign/Crop_and_Soil/manager/schedule.py

```diff
@@ -0,0 +1,171 @@
+        self.pattern_repeat = pattern_repeat
+
+    def _validate_pattern_parameters(self) -> None:
+        """
+        Checks the pattern skip and repeat parameters, if they are not correct raises errors.
+
+        Raises
+        ------
+        ValueError
+            If the skip is < 0.
+        ValueError
+            If the repeat is < 0.
+
+        """
+        if self.pattern_skip < 0:
+            raise ValueError(f"'{self.name}': expected pattern skip to be >= 0, received '{self.pattern_skip}'.")
+        if self.pattern_repeat < 0:
+            raise ValueError(f"'{self.name}': expected pattern repeat to be >= 0, received '{self.pattern_repeat}'.")
+
+    @staticmethod
+    def _elongate_list(list_to_elongate: List[Any], reference_list_length: int) -> List[Any]:
+        """
+        Takes a list and lengthens it to match the length of the reference list, if the original length was 1.
+
+        Parameters
+        ----------
+        list_to_elongate : List[Any]
+            List to be extended if its length is 1.
+        reference_list_length : int
+            Length of that the list should be extended to, if it its original length is 1.
+
+        Returns
+        -------
+        List[Any]
+            The elongated list.
+
+        Notes
+        -----
+        In the context of Schedule-descendant classes, the reference list length will always be the length of the years
+        list.
+
+        """
+        if len(list_to_elongate) != 1:
+            return list_to_elongate
+        elongated_list = list_to_elongate * reference_list_length
+        return elongated_list
+
+    @staticmethod
+    def _validate_days(days: List[int]) -> bool:
+        """
+        Checks that all values passed for days are in the correct range.
+
+        Parameters
+        ----------
+        days : List[int]
+            Julian day(s) in which this event will occur.
+
+        Returns
+        -------
+        bool
+            True if all days are valid.
+
+        Notes
+        -----
+        A day is 'valid' if it in the range [1, 366].
+
+        """
+        return all(0 < day <= 366 for day in days)
+
+    @staticmethod
+    def _validate_years(years: List[int]) -> bool:
+        """
+        Checks that all years passed are valid and ordered.
+
+        Parameters
+        ----------
+        years : List[int]
+
+        Returns
+        -------
+        bool
+            True if years are valid and ordered, False if not.
+
+        Notes
+        -----
+        A list of years is valid if every year is > 0, and the list of years does not descend at all.
+
+        """
+        return all(0 < years[index] <= years[index + 1] for index in range(0, len(years) - 1))
+
+    @staticmethod
+    def _repeat_pattern(pattern: List[int], skip: int = 0, repeat: int = 0) -> List[int]:
+        """
+        Takes a pattern of numbers and repeats it a specified number of times, skipping over specified gaps between
+        repetitions.
```

Comment:
```suggestion
        """
        Takes a pattern of differences between the list numbers and repeats it a specified number of times, skipping over specified gaps between
        repetitions.
```
or something similar to give more details about the pattern. Also, I'm not sure if this is redundant, but maybe maybe add a note to indicated if the skip is inclusive or not.

Reply:
Good suggestion